Bolts/src/bolts/CancellationTokenSource.java

```diff
@@ -19,36 +23,112 @@
+  private final Object lock = new Object();
+  private final List<CancellationTokenRegistration> registrations = new ArrayList<>();
+  private boolean cancellationRequested;
+  private boolean closed;
 
   /**
    * Create a new {@code CancellationTokenSource}.
    */
   public CancellationTokenSource() {
-    token = new CancellationToken();
   }
 
   /**
    * @return {@code true} if cancellation has been requested for this {@code CancellationTokenSource}.
    */
   public boolean isCancellationRequested() {
-    return token.isCancellationRequested();
+    synchronized (lock) {
+      throwIfClosed();
+      return cancellationRequested;
+    }
   }
 
   /**
    * @return the token that can be passed to asynchronous method to control cancellation.
    */
   public CancellationToken getToken() {
-    return token;
+    synchronized (lock) {
+      throwIfClosed();
+      return new CancellationToken(this);
+    }
   }
 
   /**
    * Cancels the token if it has not already been cancelled.
    */
   public void cancel() {
-    token.tryCancel();
+    List<CancellationTokenRegistration> registrations;
+    synchronized (lock) {
+      throwIfClosed();
+      if (cancellationRequested) {
+        return;
+      }
+      cancellationRequested = true;
+      registrations = new ArrayList<>(this.registrations);
+      this.registrations.clear();
+    }
+    notifyListeners(registrations);
+  }
+
+  /**
+   * Releases all resources associated with this {@code CancellationTokenSource}.
+   */
+  @Override
+  public void close() {
+    synchronized (lock) {
+      if (closed) {
+        return;
+      }
+      for (CancellationTokenRegistration registration : registrations) {
+        registration.close();
+      }
+      registrations.clear();
+      closed = true;
+    }
+  }
+
+  /* package */ CancellationTokenRegistration register(Runnable action) {
+    CancellationTokenRegistration ctr;
+    synchronized (lock) {
+      throwIfClosed();
+
+      ctr = new CancellationTokenRegistration(this, action);
+      if (cancellationRequested) {
+        action.run();
+      } else {
+        registrations.add(ctr);
+      }
+    }
+    return ctr;
+  }
+
+  /**
+   * @throws CancellationException if this token has had cancellation requested.
+   * May be used to stop execution of a thread or runnable.
+   */
+  /* package */ void throwIfCancellationRequested() throws CancellationException {
+    synchronized (lock) {
+      throwIfClosed();
+      if (cancellationRequested) {
+        throw new CancellationException();
+      }
+    }
+  }
+
+  /* package */ void unregister(CancellationTokenRegistration registration) {
+    synchronized (lock) {
+      throwIfClosed();
+      registrations.remove(registration);
+    }
+  }
+
+  private void notifyListeners(List<CancellationTokenRegistration> registrations) {
+    for (CancellationTokenRegistration registration : registrations) {
+      registration.runAction();
+    }
```

Comment:
Technically the notifyListeners method itself is thread safe since it doesn't access any state
But I will add a note saying the notification will not be synchronized with other events